### ISSUE TITLE
why AssertionError when training a language model?

### ISSUE THREAD

**Issue by hischen** [pytorch/fairseq#1456]
When I use fairseq to train a language model as follow:
https://github.com/pytorch/fairseq/tree/master/examples/language_model（Training a transformer language model with the CLI tools，2) Train a language model）
fairseq-train --task language_modeling \
    data-bin/wikitext-103 \
    --save-dir checkpoints/transformer_wikitext-103 \
    --arch transformer_lm_wiki103 \
    --max-update 286000 --max-lr 1.0 --t-mult 2 --lr-period-updates 270000 --lr-scheduler cosine --lr-shrink 0.75 \
    --warmup-updates 16000 --warmup-init-lr 1e-07 --min-lr 1e-09 --optimizer nag --lr 0.0001 --clip-norm 0.1 \
    --criterion adaptive_loss --max-tokens 3072 --update-freq 3 --tokens-per-sample 3072 --seed 1 \
    --sample-break-mode none --skip-invalid-size-inputs-valid-test --ddp-backend=no_c10d

I got a Assert error :
assert hasattr(model.decoder, 'adaptive_softmax') and model.decoder.adaptive_softmax is not None
AssertionError

why dose it happen,is that  I use the adaptive_loss in the incorrect way ?pls help me

**Comment by lematt1991** [comment]
Can you try adding `--adaptive-softmax-cutoff "20000,60000"`

**Comment by lematt1991** [comment]
Actually try this:

```
fairseq-train data-bin/wikitext-103 --save-dir checkpoints/transformer_wikitext-103 \
    --adaptive-input --adaptive-input-cutoff "20000,60000" --adaptive-input-factor 4 \
    --adaptive-softmax-cutoff "20000,60000" --adaptive-softmax-dropout 0.2 \
    --adaptive-softmax-factor 4 --arch "transformer_lm_gbw" --attention-dropout 0.1 \
    --char-embedder-ffn-size 0 --char-embedder-highway-layers 2 \
    --character-embedding-dim 4 --character-filters \
    "[(1, 64), (2, 128), (3, 192), (4, 256), (5, 256), (6, 256), (7, 256)]" \
    --clip-norm 0.1 --criterion "adaptive_loss" --decoder-attention-heads 8 \
    --decoder-embed-dim 1024 --decoder-ffn-embed-dim 4096 --decoder-input-dim 1024 \
    --decoder-layers 16 --decoder-normalize-before --decoder-output-dim 1024 --dropout 0.3 \
    --fp16 --keep-interval-updates 1 --lr 0.0001 --lr-period-updates 270000 --lr-scheduler \
    "cosine" --lr-shrink 0.75 --max-lr 1 --max-source-positions 3072 --max-target-positions \
    3072 --max-tokens 3072 --max-update 286000 --min-loss-scale 0.0001 --min-lr 1e-09 \
    --momentum 0.99 --no-epoch-checkpoints --optimizer "nag" --relu-dropout 0.1 \
    --save-interval 1 --save-interval-updates 270000 \
    --skip-invalid-size-inputs-valid-test --t-mult 2 --task "language_modeling" \
    --tie-adaptive-weights --tokens-per-sample 3072 --update-freq 3 --warmup-init-lr 1e-07 \
    --warmup-updates 16000 --weight-decay 0
```

**Comment by myleott** [comment]
@hischen your command works for me. Can you also try rerunning your command on master? Or installed the lastest fairseq release (0.9.0)?

**Comment by GeorgeS2019** [comment]
@lematt1991  Thanks for sharing the command. A good source to learn how to get close to the expected BLEU.

**Comment by miradel51** [comment]
It might be reasonable for setting the args --arch  "transformer_lm_wiki103"  and if it still reports the error then u can change the "cutoff" in the "transformer_lm.py" file.

**Comment by myleott** [comment]
Looks like this is resolved. FYI, we've also added a simpler transformer LM example config (which doesn't use adaptive inputs) to the README, which you might try: https://github.com/pytorch/fairseq/tree/master/examples/language_model#2-train-a-language-model
